### PR TITLE
lazy move to const generics

### PR DESCRIPTION
**Pull request by mriise**
What this PR does

- removes generic array
- implements Default for things that needed it by making an empty array `[0u8; SIZE]`

What it doesn't do:

- build
- modify proc macro
- fix SHA macro_rules! 
- some other things that need to be squashed to even build.

### CODE REVIEW

File: src/hasher.rs [hunk 1]
```diff
@@ -31,15 +19,15 @@ pub trait Digest<S: Size>:
 {
     /// Size of the digest.
     fn size(&self) -> u8 {
-        S::to_u8()
+        S as u8
```
[the remark this answers is not in the document]
Comment (mriise) [reply src/hasher.rs:22]: i guess the real limitation should be the limitation of the hash functions, so 2^64-1 bytes for some Blake2 and 3.

[the remark this answers is not in the document]
Comment (vmx) [reply src/hasher.rs:22]: Yes it is an artificial limitation. It's 256 bytes to support relatively large identity hashes. Though I would keep it like that, you don't really want to stack allocate large amounts which would lead to a stack overflow. If you need larger hashes, this library might not be a good fit :)

File: src/multihash.rs [hunk 2]
```diff
@@ -1,22 +1,19 @@
...
-use generic_array::{ArrayLength, GenericArray};
+use std::usize;
 
 /// Trait that implements hashing.
 ///
 /// It is usually implemented by a custom code table enum that derives the [`Multihash` derive].
 ///
 /// [`Multihash` derive]: crate::derive
-pub trait MultihashDigest:
+pub trait MultihashDigest<const SIZE: usize>:
```
[the remark this answers is not in the document]
Comment (vmx) [reply src/multihash.rs:14]: I find `SIZE` quite nice as it communicates well what it is about, but I don't have a strong preference. So if `S` feels better, go for `S`.

File: src/hasher_impl.rs [hunk 3]
```diff
@@ -347,15 +348,21 @@ pub mod identity {
...
     }
 
-    impl<S: Size> StatefulHasher for IdentityHasher<S> {
-        type Size = S;
-        type Digest = IdentityDigest<Self::Size>;
+    impl<const S: usize> Default for IdentityHasher<S> {
+        fn default() -> Self {
+            Self {i: 0, bytes: [0u8; S]}
+        }
+    }
```
Comment (dvc94ch) [inline src/hasher_impl.rs:361]: Why doesn't derive work?

File: src/hasher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dvc94ch) [reply src/hasher.rs:22]: Since S is mostly used to index into arrays, usize is probably more convenient. But I totally agree that users should use reasonable values like 32 or 64

**Comment by mriise** [comment]
sorry, had this up on my phone in my pocket which caused the closing and opening. Almost posted a chunk of garbled text as a comment too...

File: src/hasher_impl.rs (hunk 3 above)
Comment (mriise) [reply src/hasher_impl.rs:361]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d19c29c843a9a840ab6f40f831c53150

We might be able to implement Default for [u8; S] but I haven't come across the right way to do it though.

File: src/hasher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mriise) [reply src/hasher.rs:22]: I think moving to usize would be a generally OK move, as we will be providing defaults of reasonable values for general use. But if people want to dig a bit deeper and be unreasonable for their own reasons, that can be up to them.

**Comment by mriise** [comment]
While stepping through parsing I noticed that to use `hash<64>` in the macro we need to be able to evaluate if `alloc_size > SIZE(of hash)`, which since the support syn has for const generic `syn::GenericArgument::Const()` gives `&syn:Expr` which from what I can tell cant be evaluated in time to be used for our purposes. 

As for `alloc_size` in the macro, I moved things to `syn::LitInt` so it can easily be parsed into a `usize`.

I honestly am not sure where to go from here, It could be something I don't know or it might just be that `syn` doesn't have support for what we need to do. Honestly this seems to be leaning a bit into `const fn`s that might support what we need, but again i don't know enough about macros to be sure.

**Comment by vmx** [comment]
@mriise I had a look at the derive macro part (which is indeed quite complicated, it took me a while to get into again). You can cherry-pick this commit: https://github.com/multiformats/rust-multihash/commit/19a44ffa2103bda738b49451dcfc6f80dc30718f

It only fixes the derive macro, the build still fails. I indeed just moved things to `syn::LintInt`. I hope that helps you to make further progress.

File: src/multihash.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vmx) [reply src/multihash.rs:14]: After taking a look at the actual code, I change my mind. i think `S` is better than `SIZE`.

**Comment by mriise** [comment]
Thank you so much @vmx, IDK what it was that was confusing me so much. 
I managed to fix the sha macros and now have run into problems with digests < max size. Will look into allowing different sized Digests tomorrow morning

**Comment by mriise** [comment]
Fix for above was as simple as using a different generic for the digest, and rust compiler being amazing as it is implies `S` for us so no breaking API for the most part.
```rust
  fn multihash_from_digest<'a, D, const S: usize>(digest: &'a D) -> Multihash<SIZE>
  where
      D: Digest<S>,
      Self: From<&'a D>;
```

I removed `fn size()` with an associated const as it just returned `SIZE` anyways. We could go with changing it to a const fn, but i figured the less unstable features we use the better.

I believe the last mole to whack is serde.
It has a `[T; 0]` impl for `Deserialize` & `Serialize` so we cant just do it for ourselves without a work-around as mentioned in https://github.com/serde-rs/serde/issues/1937, alternatively we could wait for https://github.com/serde-rs/serde/pull/1860 to be merged and released, and even _further_ alternative is to use https://github.com/est31/serde-big-array for its const-generic support in the meantime.

**Comment by vmx** [comment]
> Thank you so much @vmx, IDK what it was that was confusing me so much.

It is confusing. It took my quite some time to dig into this code again. The final fix looks trivial, but getting there wasn't :)


> I believe the last mole to whack is serde.

I can't really tell what's best from reading the threads, which also means I don't have a strong preference. So I suggest trying things out and see what works (best).

**Comment by mriise** [comment]
@vmx simplest option was to use `serde-big-array`, and maybe when serde merges the PR we can drop it for that.

File: src/hasher.rs [hunk 4]
```diff
@@ -29,17 +16,15 @@ pub trait Digest<S: Size>:
...
 {
-    /// Size of the digest.
-    fn size(&self) -> u8 {
-        S::to_u8()
-    }
+    /// Size of the digest. Maximum for Some of the Blake family is 2^64-1 bytes
+    const SIZE: usize = S;
 
     /// Wraps the digest bytes.
     fn wrap(digest: &[u8]) -> Result<Self, Error> {
```
Comment (mriise) [inline src/hasher.rs:23]: im wondering if should we add 
```rust
fn wrap_arr<const S: usize>(digest: [u8; S]) -> Self
```

File: src/hasher_impl.rs [hunk 5]
```diff
@@ -213,7 +205,12 @@ macro_rules! derive_hasher_sha {
 
             fn finalize(&self) -> Self::Digest {
                 use digest::Digest;
-                Self::Digest::from(self.state.clone().finalize())
+                // TODO: this extra array seems excessive to convert from a generic array
```
Comment (mriise) [inline src/hasher_impl.rs:208]: Array here is needed, but the extra variables arent.

File: src/multihash.rs [hunk 6]
```diff
@@ -43,10 +41,9 @@ pub trait MultihashDigest:
     /// println!("{:02x?}", hash);
     /// ```
     #[allow(clippy::needless_lifetimes)]
-    fn multihash_from_digest<'a, S, D>(digest: &'a D) -> Multihash<Self::AllocSize>
+    fn multihash_from_digest<'a, D, const DIGEST_SIZE: usize>(digest: &'a D) -> Multihash<SIZE>
```
Comment (mriise) [inline src/multihash.rs:44]: TODO of moving SIZE back to S

File: src/multihash.rs [hunk 7]
```diff
@@ -73,27 +70,37 @@ pub trait MultihashDigest:
 /// ```
 #[cfg_attr(feature = "serde-codec", derive(serde::Deserialize))]
 #[cfg_attr(feature = "serde-codec", derive(serde::Serialize))]
-#[cfg_attr(feature = "serde-codec", serde(bound = "S: Size"))]
```
Comment (mriise) [inline src/multihash.rs:76]: I dont think removing this has any Ill effects, but im not 100% sure.

File: src/multihash.rs [hunk 8]
```diff
@@ -191,26 +198,23 @@ impl parity_scale_codec::Encode for Multihash<crate::U32> {
 }
 
 #[cfg(feature = "scale-codec")]
-impl parity_scale_codec::EncodeLike for Multihash<crate::U32> {}
+impl parity_scale_codec::EncodeLike for Multihash<32> {}
```
Comment (mriise) [inline src/multihash.rs:201]: do we make this
 ```rust
impl<const S: uisze> parity_scale_codec::EncodeLike for Multihash<S>
```
or does is that now how that works

**Comment by mriise** [comment]
did not mean to force push that...

**Comment by mriise** [comment]
build fails because of ```Warning: can't set `format_code_in_doc_comments = true`, unstable features are only available in nightly channel.```

File: src/hasher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mriise) [reply src/hasher.rs]: I take this back, u16 would probably be the limit for stack allocation, but at that point its probably smart to stay at u8 as there aren't many hashes that support more than 64 bytes.

I may open a PR that allows heap allocated digests in the future, as I dislike the limitation.

File: src/multihash.rs (hunk 8 above)
Comment (mriise) [reply src/multihash.rs:201]: it is how that works for v >2.1.1, but does require some modifications

File: src/hasher.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vmx) [reply src/hasher.rs]: > I may open a PR that allows heap allocated digests in the future, as I dislike the limitation.

That also something that people might want if they use identity hashes. I'm not a fan of having large identity hashes (I think identity hashes should only be used if the data is smaller than the hash you use), but that use case exists in the IPLD space. Currently, if someone would request it, I'd say "this is out of scope for this library". Though perhaps there is a nice way to support both things, stack and heap allocations (I haven't put thought into it yet).

@mriise In case you're thinking about it, please also keep the rust-cid in mind. It's not a requirement that it works with rust-cid, I just want to mention it as it might change how it might be implemented (again, perhaps there should be a separate CID library that works with heap allocations, who knows :)

[the remark this answers is not in the document]
Comment (mriise) [reply src/hasher.rs]: A separate crate sounds probably nicer just in terms of crate complexity

**Comment by mriise** [comment]
While it would be nice to have more baked in safeguards to prevent people from making excessively large hashes, there is currently no way to do this:
```rust
struct Foo<const N: u16> ([u8; N as usize]);
```
we could eventually do a check with `#![feature(const_evaluatable_checked)]` to make sure it is under a baked in number (probably 4096 as thats the largest hash i can reasonably think to be used).

File: README.md [hunk 9]
```diff
@@ -74,7 +76,7 @@ fn main() {
 * `SHA2-512`
 * `SHA3`/`Keccak`
 * `Blake2b-256`/`Blake2b-512`/`Blake2s-128`/`Blake2s-256`
-* `Blake3`
+* `Blake3`(256 only)
```
[the remark this answers is not in the document]
Comment (dvc94ch) [reply README.md:79]: yes, looks like that is the case, but looks trivial to implement. Maybe instead of changing the README we can change the implementation as it only requires one additional line of code and one change?

File: examples/custom_table.rs [hunk 10]
```diff
@@ -1,19 +1,16 @@
...
 };
 
 // You can implement a custom hasher. This is a SHA2 256-bit hasher that returns a hash that is
 // truncated to 160 bits.
 #[derive(Default, Debug)]
 pub struct Sha2_256Truncated20(Sha2_256);
-impl StatefulHasher for Sha2_256Truncated20 {
-    type Size = U20;
-    type Digest = Sha2Digest<Self::Size>;
+impl StatefulHasher<20> for Sha2_256Truncated20 {
```
Comment (dvc94ch) [inline examples/custom_table.rs:12]: this should still be an associated constant

Comment (dvc94ch) [reply examples/custom_table.rs:12]: what's `Self::SIZE`?

**Review by dvc94ch — COMMENTED** [review]
@vmx I think this answers your question

File: examples/custom_table.rs (hunk 10 above)
Comment (mriise) [reply examples/custom_table.rs:12]: `Self::SIZE` replaced `size()` since while it could be a `const fn` it made more sense (in my mind) to make it a public associated constant. 

It would make sense to move back to `const fn size()` in the case of having `Size` be an associated constant without being repetitive.

Comment (mriise) [reply examples/custom_table.rs:12]: It seems using an assosicated const as a paramater is still part of the incomplete `const_generics` feature in nightly, though with enough feature flags we can make it look something like this 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=30eed71199c4aecc0af8a0a893685084

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mriise) [reply README.md:79]: I suppose the wording should be changed? What is meant here is that multihash table currently only can understand blake3-256. People can still create larger hashes with `Blake3Hasher<S>` but everything to deal with the table will be assumed to be 256. This can be fixed by adding another case to the proc macro, but I figured another PR to do that change would be more appropriate.

File: src/hasher_impl.rs [hunk 11]
```diff
@@ -1,35 +1,38 @@
...
+use crate::hasher::{Digest, StatefulHasher};
 use core::convert::TryFrom;
-use generic_array::GenericArray;
 
 macro_rules! derive_digest {
     ($name:ident) => {
         /// Multihash digest.
-        #[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
-        pub struct $name<S: Size>(GenericArray<u8, S>);
+        #[derive(Clone, Debug, Eq, Hash, PartialEq)]
```
[the remark this answers is not in the document]
Comment (mriise) [reply src/hasher_impl.rs:8]: did you mean something else besides what's done on line 11?

File: src/multihash.rs [hunk 12]
```diff
@@ -73,27 +70,37 @@ pub trait MultihashDigest:
 /// ```
 #[cfg_attr(feature = "serde-codec", derive(serde::Deserialize))]
 #[cfg_attr(feature = "serde-codec", derive(serde::Serialize))]
-#[cfg_attr(feature = "serde-codec", serde(bound = "S: Size"))]
-#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
-pub struct Multihash<S: Size> {
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
```
[the remark this answers is not in the document]
Comment (mriise) [reply src/multihash.rs:73]: same as above, line 86

File: src/hasher_impl.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (mriise) [reply src/hasher_impl.rs]: Thanks for pointing this out though, I realized I forgot to add back in `Copy` for this.

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (vmx) [reply README.md:79]: As I understand the current README is: "you can create your own table with these hash functions". So just saying "Blake3" sounds right to me. Then probably it should also be changes to just saying "Blake2s" and "Blake2b" for consistency.

Though I tried adding a Blake3 to a custom table and it didn't work for me. I guess adding a blake3 hash with a different size to the `custom_table` example would be good (though this would be a separate PR and is not related to this change).

File: examples/custom_table.rs (hunk 10 above)
Comment (vmx) [reply examples/custom_table.rs:12]: Not being able to have it as associate constant in current stable Rust sucks :-/

File: src/hasher_impl.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (vmx) [reply src/hasher_impl.rs]: Sorry, I missed line 11. That is what I was asking for.

File: src/multihash.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (vmx) [reply src/multihash.rs]: Same as above, I missed that. What I was asking for already existed.

File: src/hasher.rs [hunk 13]
```diff
@@ -41,7 +41,7 @@ pub trait Digest<const S: usize>:
         use unsigned_varint::io::read_u64;
 
         let size = read_u64(&mut r)?;
-        if size > S as u64 || size > u8::max_value() as u64 {
+        if size > S as u64 || size > u16::MAX as u64 {
```
Comment (vmx) [inline src/hasher.rs:44]: Do we really want to increase this? Isn't 256 bytes (2048 bit) enough for hashes. I know it limits the identity hash, but even for that it's probably good enough. I also haven't heard any complaints about the size being too small. I'd keep it to 256 bytes.

(It's the same for other occurences)

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (dvc94ch) [reply README.md:79]: The blake3 hasher will indeed panic when requesting anything other than a 32 bytes digest.

[the remark this answers is not in the document]
Comment (mriise) [reply README.md:79]: I think keeping blake2b/s-xxx should be kept the same for the sake that they have different codes for each size, while Blake3 has a single code for every hash it produces, just like the sha family. Would separating the supported algorithms from the 
list of default hashers be a better way of conveying this?

File: examples/custom_table.rs (hunk 10 above)
Comment (dvc94ch) [reply examples/custom_table.rs:12]: are we in a hurry to merge this? it's a usability regression, and comparing it to the benefits, maybe we should hold off until more of const-generics is stabilized.

Comment (vmx) [reply examples/custom_table.rs:12]: I'm not in a hurry. Though I fear that it will take a long time until it's usable on stable.

@mriise What do you think about perhaps fixing everything up a single commit (once we are happy with the state of it) and we just keep it open? Then people who want const generic support can easily cherry-pick it themselves and it should also be easier to rebase it if we make other changes (e.g. fixing the blake3 hashing). This way all your work is not lost, still usable and hopefully find it's way in some day.

Comment (mriise) [reply examples/custom_table.rs:12]: Although it is disappointing, I would have to agree. I'll do the final bits of cleanup and request another review.

**Comment by mriise** [comment]
A different way to implement const generics for certian static sized algorithms could be [this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a0dea06b697be526a341631d05fb51d5). This should compile fine and is probably part of the incomplete part. This differs a bit more in complexity but only requires one flag instead of the three in my last suggested option.

**Comment by vmx** [comment]
`generic_associated_types` also seems a lot closer to become a stable feature that full `const_generics`.

**Comment by dvc94ch** [comment]
@mriise thanks for your work on this PR even if it takes a while for it to get merged

**Comment by mriise** [comment]
@vmx Yep! While I will have to put this on the back burner to make room for some other projects, I will open another PR that tries the other approach on nightly at some point.

@dvc94ch thanks!

**Comment by vmx** [comment]
@mriise Sure, as we identified it surely isn't time sensitive. I'm also disappointed that it cannot just land. I was warned, but I had hoped we really only need the `min_const_generics` part. Anyway, I'd like to thank you too, for spending so much time on this and such long review cycles.

**Comment by mriise** [comment]
TL;DR for anyone in the future, this PR made the change of `impl StatefulHasher` -> `impl<const S: usize> StatefulHasher<S>` by consequence of `Digest<const S: usize>` not being able to infer from an associated const (currently we use an associated typenum). As stated above, if you may cherry pick this PR if you need const generic support.

**Comment by koushiro** [comment]
@vmx Can you consider merging this PR? I really need it.

**Comment by vmx** [comment]
@koushiro As mentioned in [this comment](https://github.com/multiformats/rust-multihash/pull/116/files#r650296798), this PR makes the library harder to use. So the folks currently involved in this PR decided to wait until [`generic_associated_types`](https://github.com/rust-lang/rust/issues/44265) land in stable Rust. So it will be a long time until const generics support can land.

**Comment by mriise** [comment]
There may be an alternate approach i've been thinking of... I will play with it a bit and will let you guys know if it will work out.

**Comment by mriise** [comment]
Well, I about tried everything I could think. At this point it boils down to 3 alternatives:

1. Go back to using Generic Array only implementing const generics for a few items
2. Unstable features, you can minimally on GATs, but it is highly incomplete (I ran across a few ICEs)
3. An unholy mess of unsafe and pointer magic

The 1st option seems to be possible (I haven't checked but it should work fine), but personally I am personally a fan of the current implementation simply for how clean it ends up.

I will put together a PR soon with the Generic Array approach.

**Comment by vmx** [comment]
I'd be in favour of keeping the current API. I don't see a reason to change the API, just for moving from GenericArray to const generics. Though I've little experience on that side of things, so if there would be a good reason, please let me know.

Thanks again @mriise for looking into that.

**Comment by mriise** [comment]
There was one last thing I thought of before setting this aside again that relied heavily on const expressions to convert typenum into const generics, but implementation details are a bit fuzzy and the current implementation isn't ready for it anyway.

 So this may be a thing to look at again if full const generics gets stabalized before GATs do (it is anyone's guess at this point i think).

And no problem @vmx, this has been a good challenge for me.

**Comment by mriise** [comment]
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=819a4ceb088d68814c02726708ef0737 ?
